File: 06-lua/01/src/06_functions.lua
-- Слайд 12: Функции: объявление и вызов
-- Функции — значения первого класса (тип function)
-- Можно объявлять именованные и анонимные функции
-- Хранятся в переменных, передаются как аргументы

print("=== Именованная функция ===")
local function greet(name)
  return "Hello, " .. name
end
print(greet("Lua"))

print("=== Анонимная функция ===")
local f = function(x)
  return x * 2
end
print(f(21)) --> 42

print("=== Функция как аргумент (функция первого класса) ===")
local function apply(fn, value)
  return fn(value)
end
print(apply(f, 10)) --> 20
print(apply(function(x) return x .. "!" end, "wow")) --> wow!

File: 04-scala/01/src/06_Functions.scala
// 06. Функции и lambda

// Обычная функция.
// Возвращается значение последнего выражения, return обычно не нужен.
def square(x: Int): Int =
  x * x

// Несколько параметров.
def sum(a: Int, b: Int): Int =
  a + b

// Параметр со значением по умолчанию.
def greet(name: String = "World"): String =
  s"Hello, $name!"

@main
def functionsDemo(): Unit =
  println(s"square(5) = ${square(5)}")
  println(s"sum(2, 3) = ${sum(2, 3)}")
  println(greet())
  println(greet("Scala"))

  // Функцию можно хранить в переменной.
  val multiplyByTwo = (x: Int) => x * 2
  println(s"multiplyByTwo(7) = ${multiplyByTwo(7)}")

  // И передавать другой функции.
  val numbers = List(1, 2, 3)
  val result = numbers.map(multiplyByTwo)
  println(s"map(function): $result")
